File: eval-examples/Wasabi/EVAL_WASABI.m
%% EVAL WASABI 

%% 0 Build up Filename, Structure, Paths

%get the user specific pulseq path
pulseq_struct=what('pulseq-cest-library');                  % look in Matlabpath if Folder is already included
pulseq_path=pulseq_struct.path;
seq_filename='WASABI_3T_001_3p7uT_1block_5ms.seq';
seq_file_folder_path= [pulseq_path filesep 'seq-library' filesep extractBefore(seq_filename, '.seq')];
seq_file_path= [seq_file_folder_path filesep seq_filename];


% Go to seq file
seq = SequenceSBB(getScannerLimits());
gamma_hz  = seq.sys.gamma*1e-6;       

%% 2)  read data from measurement (dicom)
dcmpath=uigetdir('','Go to DICOM Directory'); cd(dcmpath)

% Question if seq file is still the same
question = input('Are the DICOM Files acquired with the same Protocoll Parameters from the PulseqCEST Library? [y/n]','s');
if strcmpi(question, 'y')
    % 2a) get the seq file from the library
    seq.read(seq_file_path)
    defs.offsets_ppm   = seq.definitions('offsets_ppm');
    Nmeas=numel(defs.offsets_ppm);
else
    % 2b) search your seqfile and resimulate the data, if there is not an
    % "M_z" textfile apparent
    [seqfile, seqpath]=uigetfile('','');
    seq.read(fullfile(seqpath,seqfile));
    defs.offsets_ppm   = seq.definitions('offsets_ppm');
    Nmeas=numel(defs.offsets_ppm);

    % 2c) check if M_z.txt is there, if not resimulate
    try 
        M_z_sim = load([seqpath extractBefore(seqfile, '.seq') '.txt']);
    catch
        disp('Simulation of M_z as a text file not found, doing re-simulation')
        M_z_sim = simulate_pulseqcest(fullfile(seqpath,seqfile),[pulseq_path filesep 'sim-library' filesep 'WM_3T_default_7pool_bmsim.yaml']);
        M_z_sim=M_z_sim'; %transpose M_z
        M_z_sim_MTRasym=M_z_sim(end:-1:2)-M_z_sim(2:end);%M0 offset is apparent
    end
end

cd(dcmpath)
collection = dicomCollection(fullfile(dcmpath));

V= dicomreadVolume(collection); 
sz=size(V); 
V=reshape(V,[sz(1) sz(2) Nmeas sz(4)/Nmeas ]); 
V= permute(V,[1 2 4 3]); size(V)

figure;
dcm_image = imagesc(squeeze(V(:, :, 6, 18)));colormap gray;colorbar

Segment=squeeze(V(:, :, :, 1)) > 100;
Segment_resh=reshape(Segment, size(Segment,1)*size(Segment,2)*size(Segment,3),1)';    % size Segment: 112x92x12

%Vectorize
V_M_z=double(permute(V,[4 1 2 3]));       % Changes from 112x92x12x32 to 32x112x92x12
sz=size(V_M_z);
maskInd=1:sz(2)*sz(3)*sz(4);
M_z=V_M_z(:,maskInd);

% Normalization
Z_wasabi=double(M_z(2:end,:))./double(M_z(1,:));


% Fitting Parameters 
B1=seq.definitions('B1');
freq=seq.definitions('FREQ');
w=defs.offsets_ppm(2:end);
t_p=seq.definitions('tp');  
fitoptions=[1E-04, 1E-15, 1E-10, 1E-4, 1E-06];
iterations = 100;
modelnum  = 021031;
c=1;
d=2;


% Compute the analytic function of WASABI curve
wasabi_fit_2abs = @(p,w) abs(p(3) - p(4) .* sin(atan((p(1) ./ (freq / gamma_)) ./ (w - p(2)))).^2 .* sin(sqrt((p(1) ./ (freq / gamma_)).^2 + (w - p(2)).^2) .* freq .* (2 * pi) .* t_p / 2).^2);
Z=Z_wasabi;

% Adapt the function for use with lsqcurvefit
% This anonymous function captures B1, freq, gamma_, and t_p from the outer scope
%wasabi_fit_function = @(params, offset) wasabi_fit_2abs(params, offset);

% Perform the WASABI fit
dB0_stack = zeros(1, size(Z, 2));
rB1_stack = zeros(1, size(Z, 2));
Z_fit = zeros(size(Z, 2), numel(w));

for ii = 1:size(Z, 2)
    if all(isfinite(Z(:, ii))) && Segment_resh(ii)==1
        p0 = [3.7, -0.1, 1, 2];                              % initial Starting values 
                           
            %opts = optimset('Display','off');                        
            [p, ~] = lsqcurvefit(wasabi_fit_2abs,p0,w,(Z(:,ii))); 
            disp(p);
            rB1_stack(ii) = p(1) / B1;
            dB0_stack(ii) = p(2);
            Z_fit(ii,:) = wasabi_fit_2abs(p,w);
              try  
        catch
            disp('something went wrong');
            rB1_stack(ii) = NaN;
            dB0_stack(ii) = NaN;
            Z_fit(ii,:) = NaN;
        end
    end
end


%Vectorization Backwards
  sizes=size(V);
  B1=reshape(rB1_stack,[sizes(1) sizes(2) sizes(3)])
  B0=reshape(dB0_stack,[sizes(1) sizes(2) sizes(3)])
  Zfit=reshape(Z_fit,[sizes(1) sizes(2) sizes(3) 31]);

 % Display B1 and B0
 figure;
 subplot(1,2,1);imagesc(B0(:,:,6),[-0.2 0.2]);colorbar;title('B0 Map');
 subplot(1,2,2);imagesc(B1(:,:,6),[0.8 1.2]);colorbar;stitle('B1 Map');

File: eval-examples/Wasabi/EVAL_WASABI.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jul 19 13:45:31 2023

@author: kouemoin
"""
import os
from pathlib import Path

import numpy as np
import pydicom
import matplotlib.pyplot as plt
import pypulseq as pp
from scipy.optimize import curve_fit
from bmctool.simulate import simulate

def EVAL_WASABI(data_flag='simulation',  data_path='', bmsim_filename='WM_3T_default_7pool_bmsim.yaml',
                seq_filename='WASABI_3T_001_3p7uT_1block_5ms.seq'):

    
    # Initializations
    seq_name = Path(seq_filename)
    seq_path = Path.cwd().parent.parent / "seq-library" / seq_name.stem / seq_name
    assert seq_path.is_file(), "seq file not found"
    
    
    # Initializations
    seq = pp.Sequence()
    seq.read(seq_path)
    
    offsets = seq.get_definition('offsets_ppm')  # offset vector [ppm]
    m0_offset = seq.get_definition('M0_offset') # corresponds to the first element in the offset vector
    freq = seq.get_definition('FREQ') # frequency [MHz]
    B1 = seq.get_definition('B1') # excitation field (B1) peak amplitude [µT]
    gamma_ = 42.578 
    t_p = seq.get_definition('tp') # pulse duration [s]
    w = offsets[1:]
    Nmeas = len(offsets) # number of repetition
    
    if data_flag == 'simulation':
        # %% 2a)  read in data from simulation
        txt_path = Path.cwd().parent.parent / "seq-library" / seq_name.stem / f'M_z_{seq_name.stem}.seq.txt'
        assert txt_path.is_file(), "Simulation data file not found"
        m_z = np.loadtxt(txt_path)
        m_z = np.expand_dims(m_z, axis=1) # Convert 1D array to 2D column vector
    
    elif data_flag == 're_simulation':
   
        #  %% 2b) Re-simulate using a yaml file
        config_path = Path.cwd().parent.parent / "sim-library" / bmsim_filename
        assert config_path.is_file(), "Config file not found"
        sim = simulate(config_file=config_path, seq_file=seq_path)
        m_z = sim.get_zspec()[1]
        m_z = np.expand_dims(m_z, axis=1)     
  
    elif data_flag == 'real_data':
        
        # %% 2c) Read the WASABI DICOM file and vectorize the data volume
        if data_path == '':
            dcmpath = input('Enter the path to your DICOM directory: ')
        else:
            dcmpath = data_path

        dcmpath = Path(dcmpath)
        assert dcmpath.is_dir(), "DICOM directory not found"
           
        collection = [pydicom.dcmread(dcmpath / filename) for filename in sorted(os.listdir(dcmpath))]

        V = np.stack([dcm.pixel_array for dcm in collection]) # Extract pixel arrays
        V = np.transpose(V, (2, 1, 0)) # Transpose to correct shape
        sz = V.shape
        V = np.reshape(V, [sz[0], sz[1], Nmeas, sz[2] // Nmeas]).transpose(0, 1, 3, 2) # Reshape to 4D volume
    
        #%% vectorization 
        mask = np.squeeze(V[:,:,:,0]) > 100 # we define a mask by selecting all the elements from the first offset and by removing any single-dimensional entries from the shape of the resulting array, effectively converting it to a 2D array. And we applied a filter.
        mask_idx = np.where(mask.ravel())[0]  # we create a 1D aaray mask index
        V_m_z =  V.reshape(-1, Nmeas).T # we reshape a 4D volume into a 2D matrix and we transpose the resulting matrix
        m_z = V_m_z[:, mask_idx]
    
    #%% normalization
    M0_idx = np.where(offsets == m0_offset)[0] # we initialize M0_idx with the index of the first element in offsets vector
    
    if len(M0_idx) > 0:
        M0 = np.mean(m_z[M0_idx,:],0)  # we compute the mean along the colunm axis
        offsets = np.delete(offsets, M0_idx) # we delete the element at the indice specified by M0_idx from the offsets
        m_z = np.delete(m_z, M0_idx,axis=0)  # similar to the previous line
        Z = m_z / M0 # Normalization
    else:
        print('m0_offset not found in offset')
    
    #%% compute the analytic function of wasabi curve    
    def wasabi_fit_2abs(w,B1,offset,c,d):  #function used for fitting data to an sinusoidal function model 
           
        y = np.abs(c - d * np.sin(np.arctan((B1 / (freq / gamma_)) / (w - offset)))**2 *
                   np.sin(np.sqrt((B1 / (freq / gamma_))**2 + (w - offset)**2) * freq * (2 * np.pi) * t_p / 2)**2) # we compute the analytic function of the wasabi curve
    
        return y
    
    # %% perform the Wasabi fit 
    dB0_stack = np.zeros(Z.shape[1]) # we initialize an array dB0_stack of zeros with a length equal to the number of columns in the Z array
    rB1_stack = np.zeros(Z.shape[1]) # same as the previous line
    Z_fit = np.zeros((Z.shape[1],np.size(w))) # same as the previous line
    
    
    for ii in range(Z.shape[1]):
         if np.all(np.isfinite(Z[:, ii])): # we check if all values in the current column ii of Z are finite (i.e., not NaN or infinity)
             
             p0=([3.7, -0.1, 1, 2]) # we initialize an initial guess. These values are initial estimates for the parameters B1, offset, c and d of wasabi_fit_2abs
             try:
                  p = curve_fit(wasabi_fit_2abs, w, Z[:,ii],  p0=p0, full_output=True, method='lm') # we use curve_fit function to fit wasabi_fit_function using Levenberg-Marquardt optimization method.
                  print(p[0])
                  rB1_stack[ii]= p[0][0] / B1 # we extract rB1_stack and normalize it by dividing with B1
                  dB0_stack[ii]= p[0][1] # we extract dB0_stack 
                  Z_fit[ii,:]=wasabi_fit_2abs(w, p[0][0], p[0][1], p[0][2], p[0][3])
                  
             except Exception:
                rB1_stack[ii]= np.nan  # in case an exception is raised during the fitting process, we set that particular column to NaN (Not-a-Number)
                dB0_stack[ii]= np.nan
                Z_fit[ii,:]= np.nan
    
    # %% vectorization backward
     
    if Z.shape[1] > 1:
        
        B0_stack = np.zeros((V_m_z.shape[1]), dtype=float) # we initialize a 2D array filled with zeros and with the same shape as V_m_z
        B0_stack[mask_idx] = dB0_stack # we assign the values from dB0_stack to B0_stack
        B0_reshaped = B0_stack.reshape(V.shape[0], V.shape[1], V.shape[2]) # we reshape B0_stack into a 3D volume
        
        B1_stack = np.zeros((V_m_z.shape[1]), dtype=float)
        B1_stack[mask_idx] = rB1_stack
        B1_reshaped = B1_stack.reshape(V.shape[0], V.shape[1], V.shape[2])
        
        Z_fit_stack = np.zeros((V_m_z.shape[1],np.size(w)), dtype=float)
        Z_fit_stack[mask_idx,:] = Z_fit
        Z_fit_stack = Z_fit_stack.reshape(V.shape[0], V.shape[1],V.shape[2],np.size(w))
        
        Z_stack = np.zeros((V_m_z.shape[1],np.size(w)), dtype=float)
        Z_stack[mask_idx,:] = Z.transpose()
        Z_stack = Z_stack.reshape(V.shape[0], V.shape[1],V.shape[2],np.size(w))
        
       
    #%% Visualization 
    
    if data_flag == 'simulation' or data_flag == 're_simulation':
        
        plt.figure(figsize=(10, 4))
        plt.subplot(1, 2, 1)
        plt.plot(w, np.mean(Z, axis=1), "r.-")  # Mittelwert über Achse 1
        plt.gca().invert_xaxis()  # x-Achse umkehren
        plt.title("Mean Z-spectrum")
        plt.xlabel("Offsets (ppm)")
        plt.ylabel("Normalized Signal")
        plt.grid(True)
        plt.show()
        
       
    if data_flag == 'real_data':
        slice_of_interest = 5  # Pick slice for evaluation
        plt.figure(figsize=(10, 4))
        plt.subplot(1, 2, 1)
        plt.imshow(B0_reshaped[:, :, slice_of_interest].squeeze(), vmin=-0.2, vmax=0.2)
        plt.colorbar()
        plt.title("B0 map")

        plt.subplot(1, 2, 2)
        plt.imshow(B1_reshaped[:, :, slice_of_interest].squeeze(), vmin=0.8, vmax=1.2)
        plt.colorbar()
        plt.title("B1 map")
        plt.show()


if __name__ == "__main__":
    EVAL_WASABI(data_flag='simulation',  
                data_path='', 
                bmsim_filename='WM_3T_default_7pool_bmsim.yaml',
                seq_filename='WASABI_3T_001_3p7uT_1block_5ms.seq')
    
    
